### PR TITLE
[autoscaler][gcp] Migrate GCP config to available node types

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/util.py [hunk 1]
```diff
@@ -109,6 +109,32 @@ def validate_config(config: Dict[str, Any]) -> None:
...
+        cli_logger.warning(
+            "The `worker_nodes` field is deprecated and will be ignored."
+            "Use `available_node_types` instead.")
+    if "available_node_types" not in config:
+        cli_logger.error("`available_node_types` not specified in config")
+        raise ValueError("`available_node_types` not specified in config")
+    if "head_node_type" not in config:
+        cli_logger.error("`head_node_type` not specified in config")
+        raise ValueError("`head_node_type` not specified in config")
+
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/util.py:136]: Pasted from https://github.com/ray-project/ray/pull/15584

**Comment by DmitriGekhtman** [comment]
Verified that I'm able to launch a two-node cluster with these changes.

File: python/ray/autoscaler/_private/gcp/config.py [hunk 2]
```diff
@@ -367,23 +378,18 @@ def _configure_subnet(config, compute):
...
+    default_interfaces = [{
+        "subnetwork": default_subnet["selfLink"],
+        "accessConfigs": [{
+            "name": "External NAT",
+            "type": "ONE_TO_ONE_NAT",
+        }],
+    }]
+
+    for node_config in node_configs:
+        if "networkInterfaces" not in node_config:
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/gcp/config.py:390]: Can you add a test for this function to verify that the behavior is maintained before & after this change?

Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/gcp/config.py:390]: I'll add a test that verifies this piece of data is inserted when not all networkInterfaces fields are specified and not inserted when they are specified.

Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/gcp/config.py:390]: Hmm, I meant "check that missing networkInterface fields are filled and list subnets isn't called if everything is already configured"

**Comment by DmitriGekhtman** [comment]
okidoke, added test and removed gcp/example-full-legacy

File: python/ray/autoscaler/_private/gcp/config.py [hunk 3]
```diff
@@ -351,10 +359,13 @@ def _configure_key_pair(config, compute):
 def _configure_subnet(config, compute):
     """Pick a reasonable subnet if not specified by the config."""
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/gcp/config.py:361]: Could we actually **not** mutate the input. We can avoid that by doing a `copy.deepcopy` of the cofnig.

**Review by ijrsvt — COMMENTED** [review]
One quick comment about making `_configure_subnet` **not** mutate the input.

File: python/ray/autoscaler/_private/gcp/config.py (hunk 3 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/gcp/config.py:361]: Yes, brilliant, I'll make that change.

File: python/ray/tests/test_autoscaler_yaml.py [hunk 4]
```diff
@@ -396,6 +397,126 @@ def testAzureKeyPair(self):
...
+
+        # Track number of times list_subnets has been called.
+        list_subnets_counter = 0
+
+        def mock_list_subnets(*args):
+            nonlocal list_subnets_counter
+            list_subnets_counter += 1
+            return [{"selfLink": "link"}]
+
+        with patch("ray.autoscaler._private.gcp.config._list_subnets",
```
Comment (DmitriGekhtman) [inline python/ray/tests/test_autoscaler_yaml.py:456]: this line leads to the error below in the CI but not locally. Anyone have an idea what's going on?
```
thing = <module 'ray.autoscaler._private.gcp' from '/ray/python/ray/autoscaler/_private/gcp/__init__.py'>
--
  | comp = 'config', import_path = 'ray.autoscaler._private.gcp.config'
  |  
  | def _dot_lookup(thing, comp, import_path):
  | try:
  | >           return getattr(thing, comp)
  | E           AttributeError: module 'ray.autoscaler._private.gcp' has no attribute 'config'

```

**Comment by DmitriGekhtman** [comment]
I'm getting a module-not-found error in `test_autoscaler_yaml.py` (one of the `small` tests) due to `gcp/config.py` importing `googleapiclient`. 
Anyone have suggestions for how to deal with that (maybe add a pip install somewhere in the CI)?
cc @simon-mo

**Comment by simon-mo** [comment]
Feel free to add it in the `python/requirements.txt`

**Comment by DmitriGekhtman** [comment]
Tests fixed
@ijrsvt 
Could you take final glance and merge if it looks good?

**Review by ijrsvt — APPROVED** [review]
Looks good!

**Pull request merged**